src/blueteam/Gardener.java

```diff
@@ -137,30 +165,137 @@ boolean build(RobotType type) {
 		}
 	}
 
-	@Override void step() {
-		// TODO Lukas
+	/**
+	 * assuming were in garden so build direction is NOT random
+	 *
+	 * @param type of robot
+	 * @return if successful
+	 */
+	boolean build(RobotType type) {
+		return build(type, false);
+	}
+
+	/**
+	 * checks whether enemy Archon is nearby,
+	 * used in init phase
+	 *
+	 * @return true if is near
+	 */
+	boolean isEnemyArchonNear() {
+		RobotInfo[] nearbyRobots = rc.senseNearbyRobots();
+		for (RobotInfo robot : nearbyRobots) {
+			if (robot.getType() == RobotType.ARCHON && robot.getTeam() == enemy)
+				return true;
+		}
+		return false;
+	}
 
+	/**
+	 * Checks whether there is a lot (const) trees around
+	 *
+	 * @return true if is more than TeamConstants.GARDENER_NUM_OF_TREES_TO_BUILD_LUMBER
+	 */
+	boolean isInWoods() {
+		TreeInfo[] nearbyTrees = rc.senseNearbyTrees(TeamConstants.GARDENERS_DEFAULT_FREE_SPOT_RADIUS, Team.NEUTRAL);
+		System.out.println(nearbyTrees.length);
+		if (nearbyTrees.length > TeamConstants.GARDENER_NUM_OF_TREES_TO_BUILD_LUMBER)
+			return true;
+		else
+			return false;
+	}
+	@Override void dodge()
+	{
+
+	}
+	/**
+	 * States:
+	 * <p>
+	 * STARTING <-> ONLYSOLDIER
+	 * \
+	 * \-----> FINDING <-> LETSCHOP
+	 * \
+	 * \-----> BUILDING
+	 */
+	@Override void step() {
+		roundCounter++;
 		switch (state) {
+		case STARTING:
+			currentDir = randomDirection();
+			if (isEnemyArchonNear()) {
+				state = GardenerState.ONLYSOLDIERS;
+			} else
+				state = GardenerState.FINDING;
+			break;
 		case FINDING:
-			if (findSpot()) {
-				state = GardenerState.BUILDING;
-				System.out.println("Gardener ID: " + rc.getID() + " found place to build garden");
-			} else {
-				tryMove(randomDirection());
+			while (roundCounter < 20) {
+				if (findSpot())
+					state = GardenerState.BUILDING;
+				roundCounter++;
+				Clock.yield();
 			}
+			if (isInWoods() && lumberjackBuilded < 1) {
+				state = GardenerState.LETSCHOP;
+				lumberjackBuilded++;
+			} else if (findSpot())
+				state = GardenerState.BUILDING;
 			break;
 		case BUILDING:
-			buildGarden();
+			// try to build one tree in garden
+			if (buildGarden())
+				initTreeNum++;
+			// water garden
 			waterGarden();
-			//TODO decide according to game state whether to build lumberjacks,
-			//soldiers or tanks. Logic how many gardeners should be build should be in archon
-			double rnd = Math.random();
-			if (rnd < 0.1)
-				build(RobotType.LUMBERJACK);
-			else if (rnd < 0.9)
+			//build atleast 3 trees
+			if (initTreeNum < 3)
+				break;
```

Comment:
Value of TeamConstants.GARDENER_NUM_OF_TREES_TO_BUILD_LUMBER is really dependent on this value as well. What happens on same maps is that you see only 3 neutral trees around you ( This happens when you are close to the map edge) and you are able to build only 2 our trees. In this scenario gardener is waiting until lumberjack from other gardener comes to help him.  Lower value of this constant should help.